### ISSUE TITLE
Distinguish custom and implementation-specific support levels

### ISSUE THREAD

**Issue by jpeach** [kubernetes-sigs/gateway-api#530]
Some APIs non-code APIs are marked "Support: Implementation-specific", but the documentation only specifies core, extended and custom levels for conformance. The "implementation-specific" level should be defined as well.

ref #482

**Comment by hbagdi** [comment]
Doesn't implementation specific imply custom conformance level?

**Comment by jpeach** [comment]
> Doesn't implementation specific imply custom conformance level?
> (That might not be obvious and we might have not documented it)

Yeh I think all we need to do is a bit of docs.

We say:
```
CUSTOM features are those that are not portable and are vendor-specific. CUSTOM features will not have API types and schema except via generic extension points
```

not sure how that differs from "implementation-specific"?

**Comment by robscott** [comment]
Maybe we should just standardize on using "Implementation-Specific" instead of "Custom"? I think they are essentially the same thing, but I think "Implementation-Specific" is a bit easier to understand.

**Comment by hbagdi** [comment]
I agree here.
One reason I remember we went with "Custom" was to promote experimentation and then graduate successful adoptions to "Extended". "ImplementationSpecific" communicates the same but more directly. The difference is really fine and probably doesn't even matter.

+1 to moving everything to ImplementationSpecific.
This again will be a large change, tagging for v1alpha2.

**Comment by Miciah** [comment]
As general terms within the context of service-apis, "implementation specific" and "custom" seem to me to be distinct, independent concepts.  An implementation-specific API may be required or optional  and leaves some semantics up to the implementation, whereas a custom API is optional and leave no, some, or all details up to the implementation.  Some examples: 

* `httproutes.spec.rules[*].matches[*].path.type: RegularExpression` is an "implementation specific" API because service-apis does not specify the regexp syntax (POSIX basic, POSIX extended, PCRE, ...).  
* `httproutes.spec.rules[*].forwardTo[*].filters` is a "custom" API because it is completely optional to implement (even though the API is well defined and therefore not implementation-specific).  
*  `httproutes.spec.tls.certificateRef.kind: <resource>` for any `<resource>` other than "secrets" (and `httproutes.spec.tls.certificateRef.group: <group>` for any `<group>` other than the core APIs group) is both custom (optional) and implementation-specific (implementations supply all the semantics for how the referenced object is used).  

It does seem like it would make sense to settle on a single term where we variously use "custom" or "implementation-specific" for describing the support level though.  "Custom" makes sense, although perhaps "Optional" would be clearer.

**Comment by bowei** [comment]
As per the discussion in the meeting on 03-24:

* Three conformance levels: core, extended and custom. These define what the requirements are for the controller:
   * core: must be supported by all controllers
   * extended: must be specified in the API (i.e. this repo), but support is not required. If support exists for the feature, it must behave as specified in this API (e.g. it's portable).
   * custom: up to the implementation; API says nothing about the behavior.

Now within a given feature, there may be areas that, due to reality, cannot be fully specified/rely on implementation details. This is just an unfortunate consequence of the fact that the area being defined by the API is quite mature with many extant implemenations. In this case, we say that an aspect of a API feature is "implementation specific". In this case, it is still useful (for our users) to spec out the feature as much as possible, but we acknowledge that the behavior for some subset of the API may still vary (and that's ok).

**Comment by hbagdi** [comment]
@bowei This is included in the v0.3.0 milestone and we are getting close to release. Do you think you will be able to get to this?

**Comment by bowei** [comment]
Yes, let me get the doc PR together
